compiler/rustc_infer/src/infer/nll_relate/mod.rs

```diff
@@ -603,61 +603,52 @@ where
-            //   subtyping (i.e., `&'b u32 <: &{P} u32`).
-            let variance = std::mem::replace(&mut self.ambient_variance, ty::Variance::Covariant);
-
-            // Note: the order here is important. Create the placeholders first, otherwise
-            // we assign the wrong universe to the existential!
-            let b_replaced = self.instantiate_binder_with_placeholders(b);
-            let a_replaced = self.instantiate_binder_with_existentials(a);
-
-            self.relate(a_replaced, b_replaced)?;
-
-            self.ambient_variance = variance;
-        }
+        match self.ambient_variance {
+            ty::Variance::Covariant => {
+                // Covariance, so we want `for<..> A <: for<..> B` --
+                // therefore we compare any instantiation of A (i.e., A
+                // instantiated with existentials) against every
+                // instantiation of B (i.e., B instantiated with
+                // universals).
+
+                // Note: the order here is important. Create the placeholders first, otherwise
+                // we assign the wrong universe to the existential!
+                let b_replaced = self.instantiate_binder_with_placeholders(b);
+                let a_replaced = self.instantiate_binder_with_existentials(a);
+
+                self.relate(a_replaced, b_replaced)?;
+            }
 
-        if self.ambient_contravariance() {
-            // Contravariance, so we want `for<..> A :> for<..> B`
-            // -- therefore we compare every instantiation of A (i.e.,
-            // A instantiated with universals) against any
-            // instantiation of B (i.e., B instantiated with
-            // existentials). Opposite of above.
+            ty::Variance::Contravariant => {
+                // Contravariance, so we want `for<..> A :> for<..> B` --
+                // therefore we compare every instantiation of A (i.e., A
+                // instantiated with universals) against any
+                // instantiation of B (i.e., B instantiated with
+                // existentials). Opposite of above.
 
-            // Reset ambient variance to contravariance. See the
-            // covariant case above for an explanation.
-            let variance =
-                std::mem::replace(&mut self.ambient_variance, ty::Variance::Contravariant);
+                let a_replaced = self.instantiate_binder_with_placeholders(a);
+                let b_replaced = self.instantiate_binder_with_existentials(b);
 
-            let a_replaced = self.instantiate_binder_with_placeholders(a);
-            let b_replaced = self.instantiate_binder_with_existentials(b);
+                self.relate(a_replaced, b_replaced)?;
+            }
+
+            ty::Variance::Invariant => {
+                // Invariant, so we want `for<..> A == for<..> B` --
+                // therefor we want `for<..> A <: for<..> B` and
+                // `for<..> B <: for<..> A`.
```

Comment:
this doesn't read right to me, the previous behaviour was wanting `A <: B` and `B <: A` but now that's changed. This comment implies that two types being subtypes of eachother implies they're the same type but that's explicitly what this PR is moving away from.

```suggestion
                // Invariant, so we want `for<..> A == for<..> B` --
                // therefor we want `exists<..> A == for<..> B` and
                // `exists<..> B == for<..> A`.
```
might be better to capture that the `relate` below is done with invariance so we are not involving subtyping here. when I first read this code that didnt stand out to me and it was somewhat confusing how this was any different than before this PR.

Reply:
:facepalm: